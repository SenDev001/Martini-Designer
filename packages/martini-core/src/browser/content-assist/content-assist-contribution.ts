import { injectable, multiInject } from "inversify";
import { ContentAssist } from "./content-assist";

export const ContentAssistContribution = Symbol("ContentAssistContribution");

export interface ContentAssistContribution {
    registerProviders(id: string, contentAssist: ContentAssist): void;
}

@injectable()
export class ContentAssistContributionManager implements ContentAssistContribution {
    @multiInject(ContentAssistContribution)
    private readonly contributions: ContentAssistContribution[];

    registerProviders(id: string, contentAssist: ContentAssist): void {
        this.contributions.forEach(contrib => contrib.registerProviders(id, contentAssist));
    }
}
