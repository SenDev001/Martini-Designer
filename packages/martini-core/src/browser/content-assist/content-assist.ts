import { MaybePromise } from "@theia/core";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import { OverlayIconProps } from "../components/icon-overlay";
import { HistoryManager } from "../history/history-manager";

export interface ContentAssist {
    enabled: boolean;
    show(props: ContentAssistProps): void;
    proposalProviders: ContentProposalProvider[];
    dispose(): void;
}

export interface ContentAssistProps {
    searchPlaceholder?: string;
    search?: string;
    proposalProviderOverrides?: ContentProposalProvider[];
}

export type Selection = any[];

export interface ContentAssistQuery {
    search: string;
    target: any;
    selection: Selection;
}

export interface ContentProposalProvider {
    getProposals(query: ContentAssistQuery, acceptor: (proposals: ContentProposal[]) => void): void;
}

export namespace ContentProposalProvider {
    export function make(provider: () => ContentProposal[]): ContentProposalProvider {
        return {
            getProposals: (_, acceptor) => acceptor(provider())
        };
    }
}

export interface ContentProposal {
    label: React.ReactNode;
    iconClass?: string;
    overlayIcons?: OverlayIconProps[];
    information?: React.ReactNode;
    sortValue?: string;
    relevance?: number;
    apply: (context: ContentAssistContext) => MaybePromise<Selection | undefined>;
    postApply?: (selection: Selection | undefined) => void;
    getApplyModes?: (selection: Selection, target: any) => ApplyMode[];
}

export namespace ContentProposal {
    export function noop(init: Omit<ContentProposal, "apply" | "postApply" | "getApplyModes">): ContentProposal {
        return {
            ...init,
            apply: () => undefined
        };
    }

    export const none = ContentProposal.noop({
        label: messages.none,
        iconClass: "martini-tree-icon martini-disabled-icon"
    });
}

export interface TemplateProposal {
    next(
        context: ContentAssistContext,
        proposalProviders: ContentProposalProvider[],
        proposal: ContentProposal
    ): TemplateContext;
    postTemplateApply?(): void;
}

export namespace TemplateProposal {
    export function is(object: any): object is TemplateProposal {
        return !!object && "next" in object;
    }
}

export interface TemplateContext {
    proposalProviders: ContentProposalProvider[];
    selection?: Selection;
    message?: string;
}

export interface ApplyMode {
    name: string;
    metaKey?: boolean;
    altKey?: boolean;
    ctrlKey?: boolean;
    shiftKey?: boolean;
}

export namespace ApplyMode {
    export function noKey(applyMode: ApplyMode): boolean {
        return !applyMode.metaKey && !applyMode.altKey && !applyMode.ctrlKey && !applyMode.shiftKey;
    }

    export function matches(applyMode: ApplyMode, modifiers: Omit<ApplyMode, "name">): boolean {
        return (
            modifiers.metaKey === (applyMode.metaKey || false) &&
            modifiers.altKey === (applyMode.altKey || false) &&
            modifiers.shiftKey === (applyMode.shiftKey || false) &&
            modifiers.ctrlKey === (applyMode.ctrlKey || false)
        );
    }
}

export interface ContentAssistContext {
    selection: Selection;
    target: any;
    historyManager: HistoryManager;
    metaKey: boolean;
    altKey: boolean;
    ctrlKey: boolean;
    shiftKey: boolean;
}

export interface ContentAssistAdapter {
    getInsertionBounds(): InsertionBounds;
    getTarget(): any;
    getSelection(): Selection;
    setSelection(selection: Selection): Promise<void>;
}

export interface InsertionBounds {
    x: number;
    y: number;
    width: number;
    height: number;
}

export abstract class BaseContentProposal implements ContentProposal {
    label: React.ReactNode;
    iconClass: string | undefined;
    overlayIcons: OverlayIconProps[] | undefined;
    category: string | undefined;
    sortValue: string | undefined;
    relevance: number | undefined;
    abstract apply(context: ContentAssistContext): MaybePromise<Selection | undefined>;
}
