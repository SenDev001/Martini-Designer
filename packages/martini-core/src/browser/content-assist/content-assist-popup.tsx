import { CommandRegistry as PhosphorCommandRegistry } from "@phosphor/commands";
import { Menu } from "@phosphor/widgets";
import { debounce } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import { CellProps } from "react-table";
import styled from "styled-components";
import { Icon, OverlayIcon } from "../components/icon-overlay";
import { Loader } from "../components/loader";
import { CellClickEvent, Table, TableColumn } from "../components/table";
import { ApplyMode, ContentProposal, ContentProposalProvider, Selection } from "./content-assist";

export interface ContentAssistPopupProps {
    proposalProviders: ContentProposalProvider[];
    context: {
        selection: Selection;
        target: any;
    };
    defaultSearch?: string;
    searchPlaceholder?: string;
    onSelectionChange?: (proposal?: ContentProposal) => void;
    onApply: (e: ApplyEvent) => void;
    onCancel: () => void;
}

export interface ApplyEvent {
    proposal: ContentProposal,
    altKey: boolean;
    ctrlKey: boolean;
    metaKey: boolean;
    shiftKey: boolean;
}

interface ContentAssistPopupState {
    search: string;
    proposals: ContentProposal[];
    selectedRow: number;
    applyModes?: ApplyMode[];
    activeApplyMode?: ApplyMode;
    loading: boolean;
}

const Styles = styled.div`
    display: grid;
    grid-row-gap: 0;
    grid-template-rows: max-content 1fr max-content;
    width: 100%;
    height: 100%;
    background-color: var(--theia-editorWidget-background);
    border: 1px solid var(--theia-contrastBorder);
    box-shadow: 0 0px 8px var(--theia-widget-shadow);

    .table {
        .tbody {
            width: 100%;
            height: 100%;

            .td {
                width: 100% !important;
            }
        }
    }

    .no-matches-label {
        text-align: center;
        padding: 10px;
    }

    .footer {
        padding-left: 4px;
        padding-right: 4px;
        font-size: 11px;
        display: grid;
        align-items: center;
        grid-template-columns: 1fr max-content;
        min-height: 20px;
        background-color: var(--theia-titleBar-activeBackground);

        .settings {
            cursor: pointer;
        }
    }
`;

const ContentProposalCell = styled.div`
    display: grid;
    grid-template-columns: max-content 1fr;
    align-items: center;
    padding-left: 2px;
    font-size: 12px;

    > * {
        text-overflow: ellipsis;
        white-space: nowrap;
        overflow-x: hidden;
    }
`;

export class ContentAssistPopup extends React.Component<ContentAssistPopupProps, ContentAssistPopupState> {
    private searchSession: number = 0;
    private readonly column: TableColumn = {
        accessor: "label",
        disableResizing: true,
        Cell: props => this.renderProposal(props)
    };
    private readonly currentModifiers = {
        metaKey: false,
        altKey: false,
        shiftKey: false,
        ctrlKey: false
    };
    private sortByName = false;
    private searchInput: HTMLInputElement | null;
    private container: HTMLDivElement | null;
    private settingsMenuOpened = false;
    lock: Promise<void> = Promise.resolve();

    constructor(props: ContentAssistPopupProps) {
        super(props);
        this.state = {
            proposals: [],
            search: this.props.defaultSearch || "",
            selectedRow: -1,
            loading: false,
        };
    }

    componentDidMount() {
        this.container?.addEventListener("focusout", e => this.handleFocusout(e));
        this.searchInput?.focus();
        this.searchInput?.select();
        this.doSearch();
    }

    componentDidUpdate(_: Readonly<ContentAssistPopupProps>, prevState: Readonly<ContentAssistPopupState>) {
        if (prevState.selectedRow !== this.state.selectedRow)
            this.props.onSelectionChange?.(this.state.proposals[this.state.selectedRow]);
    }

    componentWillUnmount() {
        this.search.cancel();
    }

    private handleSearch(e: React.ChangeEvent<HTMLInputElement>) {
        this.setState({
            ...this.state,
            search: e.target.value
        });
        this.search();
    }

    private readonly search = debounce(() => this.doSearch(), 300);

    private doSearch() {
        if (!this.props.proposalProviders.length)
            return;

        const session = this.searchSession = Date.now();
        const { selection, target } = this.props.context;
        const search = this.state.search.trim().toLowerCase();
        this.setState({
            ...this.state,
            proposals: [],
            selectedRow: -1,
            loading: true
        });
        let acceptCounter = 0;
        const acceptor = async (proposals: ContentProposal[]) => {
            if (session !== this.searchSession)
                return;

            await this.lock;

            this.lock = new Promise(resolve => {
                acceptCounter++;
                const combinedProposals = [...this.state.proposals, ...proposals];
                let sortedProposals: ContentProposal[];

                if (search.length !== 0 && !this.sortByName)
                    sortedProposals = combinedProposals.sort((p1, p2) => (p1.relevance || 0) - (p2.relevance || 0));
                else
                    sortedProposals = combinedProposals.sort((p1, p2) => (p1.sortValue || "").localeCompare(p2.sortValue || ""));

                const selectedRow = sortedProposals.length !== 0 ? 0 : -1;
                const applyModes = this.getApplyModes(sortedProposals[selectedRow]);
                this.setState({
                    ...this.state,
                    selectedRow,
                    proposals: sortedProposals,
                    applyModes,
                    activeApplyMode: this.matchApplyMode(applyModes),
                    loading: !this.state.loading ? false : (acceptCounter !== this.props.proposalProviders.length && proposals.length === 0)
                }, resolve);
            });
        };
        this.props.proposalProviders.forEach(provider => {
            provider.getProposals({
                search,
                target,
                selection
            }, acceptor);
        });
    }

    private getApplyModes(proposal?: ContentProposal): ApplyMode[] {
        if (!proposal || !proposal.getApplyModes)
            return [];
        const { selection, target } = this.props.context;
        return proposal.getApplyModes!(selection, target);
    }

    private matchApplyMode(applyModes: ApplyMode[]): ApplyMode | undefined {
        return applyModes.find(ap => ApplyMode.matches(ap, this.currentModifiers)) || applyModes.find(ap => ApplyMode.noKey(ap));
    }

    private handleGlobalKeyUp(e: React.KeyboardEvent) {
        this.updateModifiers(e);
        if (e.key === "Enter" && this.state.selectedRow >= 0) {
            this.props.onApply({
                proposal: this.state.proposals[this.state.selectedRow],
                altKey: e.altKey,
                shiftKey: e.shiftKey,
                ctrlKey: e.ctrlKey,
                metaKey: e.metaKey
            });
            e.preventDefault();
        }
        else if (e.key === "Escape")
            this.props.onCancel();
        else if (this.searchInput !== document.activeElement && e.key.length === 1) {
            this.setState({
                ...this.state,
                search: this.state.search + e.key
            });
            this.search();
            this.searchInput?.focus();
        }
    }

    private handleGlobalKeyDown(e: React.KeyboardEvent) {
        this.updateModifiers(e);
    }

    private updateModifiers(e: React.KeyboardEvent) {
        const { altKey, metaKey, shiftKey, ctrlKey } = this.currentModifiers;

        if (altKey === e.altKey && metaKey === e.metaKey && shiftKey === e.shiftKey && ctrlKey === e.ctrlKey)
            return;

        Object.assign(this.currentModifiers, {
            altKey: e.altKey,
            metaKey: e.metaKey,
            ctrlKey: e.ctrlKey,
            shiftKey: e.shiftKey
        });

        const activeApplyMode = this.matchApplyMode(this.state.applyModes || []);

        this.setState({
            ...this.state,
            activeApplyMode
        });
    }

    private getApplyModeLabel(applyMode: ApplyMode) {
        let label = applyMode.name + " ";

        if (applyMode.metaKey)
            label += PhosphorCommandRegistry.formatKeystroke("Accel");
        if (applyMode.shiftKey)
            label += PhosphorCommandRegistry.formatKeystroke("Shift");
        if (applyMode.altKey)
            label += PhosphorCommandRegistry.formatKeystroke("Alt");
        if (applyMode.ctrlKey)
            label += PhosphorCommandRegistry.formatKeystroke("Ctrl");

        // "+" is appended on linux/windows
        return label.replace("+", "");
    }

    private handleTextInputKeyDown(e: React.KeyboardEvent) {
        if (this.state.proposals.length !== 0 && (e.key === "ArrowUp" || e.key === "ArrowDown")) {
            let index = this.state.selectedRow;

            if (e.key === "ArrowUp")
                index--;
            else
                index++;

            if (index < 0)
                index = this.state.proposals.length - 1;
            else if (index >= this.state.proposals.length)
                index = 0;
            e.preventDefault();
            e.stopPropagation();
            this.handleSelectionChange([index]);
        }
    }

    private handleSelectionChange(selectedRows: number[]) {
        this.setState({
            ...this.state,
            selectedRow: selectedRows[0]
        });
    }

    private handleProposalDoubleClick(e: CellClickEvent) {
        this.props.onApply({
            proposal: e.value,
            altKey: e.mouseEvent.altKey,
            shiftKey: e.mouseEvent.shiftKey,
            ctrlKey: e.mouseEvent.ctrlKey,
            metaKey: e.mouseEvent.metaKey
        });
    }

    private handleSettingsClick(e: React.MouseEvent) {
        const commands = new PhosphorCommandRegistry();
        commands.addCommand("sortByName", {
            label: messages.sort_by_name,
            isToggled: () => this.sortByName,
            execute: async () => {
                this.sortByName = !this.sortByName;
                this.doSearch();
            }
        });
        const menu = new Menu({
            commands
        });
        menu.addItem({
            command: "sortByName"
        });
        menu.aboutToClose.connect(() => {
            this.settingsMenuOpened = false;
            this.searchInput?.focus();
        });
        const bounds = e.currentTarget.getBoundingClientRect();
        this.settingsMenuOpened = true;
        menu.open(bounds.left, bounds.bottom);
    }

    private handleFocusout(e: FocusEvent) {
        if (this.settingsMenuOpened) {
            e.preventDefault();
            e.stopPropagation();
        }
    }

    private renderProposal(props: React.PropsWithChildren<CellProps<any, any>>): React.ReactNode {
        const proposal = props.row.original as ContentProposal;
        return <ContentProposalCell>
            <Icon iconClass={proposal.iconClass || ""}>
                {proposal.overlayIcons &&
                    proposal.overlayIcons.map((props, i) => (<OverlayIcon key={i} {...props} />))}
            </Icon>
            {proposal.label}
        </ContentProposalCell>;
    }

    render(): React.ReactNode {
        let body: React.ReactNode;

        if (!this.state.loading && this.state.proposals.length) {
            body = <Table
                columns={[this.column]}
                data={this.state.proposals}
                showTableHeader={false}
                multiSelection={false}
                focusable={true}
                onSelectionChange={rows => this.handleSelectionChange(rows)}
                onCellDoubleClick={e => this.handleProposalDoubleClick(e)}
                selectedRows={[this.state.selectedRow]}
            />;

        }
        else if (!this.props.proposalProviders.length || (!this.state.proposals.length && !this.state.loading))
            body = <div className="no-matches-label">{messages.no_matches}</div>;
        else if (this.state.loading)
            body = <Loader style={{ position: "relative" }} />;

        return <Styles
            onKeyUp={e => this.handleGlobalKeyUp(e)}
            onKeyDown={e => this.handleGlobalKeyDown(e)}
            ref={elem => this.container = elem}
        >
            <input
                type="text"
                className="theia-input"
                placeholder={this.props.searchPlaceholder || messages.search_placeholder}
                value={this.state.search}
                onChange={e => this.handleSearch(e)}
                onKeyDown={e => this.handleTextInputKeyDown(e)}
                ref={elem => this.searchInput = elem}
            />
            {body}
            <div className="footer" onMouseDown={e => e.preventDefault()}>
                <div
                    className="apply-mode"
                    title={this.state.applyModes?.map(ap => this.getApplyModeLabel(ap)).join("\n")}>
                    {this.state.activeApplyMode && messages.add_mode(this.state.activeApplyMode.name)}
                </div>
                <i
                    className="codicon codicon-settings-gear settings"
                    title={messages.settings}
                    onClick={e => this.handleSettingsClick(e)}
                />
            </div>
        </Styles>;
    }
}
