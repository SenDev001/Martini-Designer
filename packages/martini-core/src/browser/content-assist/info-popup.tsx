import * as React from "react";
import styled from "styled-components";
import { ContentProposal } from "./content-assist";

export interface InfoPopupProps {
    proposal: ContentProposal;
}

const Styles = styled.div`
    background-color: var(--theia-editorWidget-background);
    border: 1px solid var(--theia-contrastBorder);
    box-shadow: 0 0px 8px var(--theia-widget-shadow);
    padding: var(--theia-ui-padding);
    font-size: 12px;
    overflow: auto;
`;

export const InfoPopup: React.FC<InfoPopupProps> = ({ proposal }) => {
    return <Styles tabIndex={-1}>
        {proposal.information}
    </Styles>;
};
