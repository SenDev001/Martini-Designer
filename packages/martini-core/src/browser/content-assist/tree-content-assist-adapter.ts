import { ExpandableTreeNode } from "@theia/core/lib/browser";
import { BaseTreeWidget } from "../tree/base-tree-widget";
import { ContentAssistAdapter, InsertionBounds, Selection } from "./content-assist";

export class TreeContentAssistAdapter implements ContentAssistAdapter {
    constructor(private readonly tree: BaseTreeWidget) {}

    getInsertionBounds(): InsertionBounds {
        const treeBounds = this.tree.node.getBoundingClientRect();
        const selectedNode = this.tree.model.selectedNodes[0];
        if (selectedNode) {
            const element = this.tree.node.querySelector(`[data-node-id='${selectedNode.id}'`);
            if (element) return element.parentElement!.getBoundingClientRect();
        }
        return {
            x: treeBounds.x,
            y: treeBounds.y,
            width: treeBounds.width,
            height: 20
        };
    }

    getTarget(): any {
        const firstSelected = this.tree.model.selectedElements[0];

        if (!firstSelected) return this.tree.model.rootData;

        return firstSelected;
    }

    getSelection(): Selection {
        return this.tree.model.selectedElements;
    }

    async setSelection(selection: Selection): Promise<void> {
        await this.tree.model.waitForRefresh;
        this.tree.model.selectElements(element => selection.includes(element));
        const node = this.tree.model.findNode(selection[0]);
        if (node && node.parent && ExpandableTreeNode.is(node.parent)) await this.tree.model.expandNode(node.parent);
        await this.tree.forceRender();
    }
}
