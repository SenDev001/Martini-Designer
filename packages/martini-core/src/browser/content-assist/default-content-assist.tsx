import { debounce } from "lodash";
import * as React from "react";
import * as ReactDOM from "react-dom";
import { HistoryManager } from "../history/history-manager";
import { ContentAssist, ContentAssistAdapter, ContentAssistProps, ContentProposal, ContentProposalProvider, InsertionBounds, TemplateContext, TemplateProposal } from "./content-assist";
import { ApplyEvent, ContentAssistPopup } from "./content-assist-popup";
import { InfoPopup } from "./info-popup";

export class DefaultContentAssist implements ContentAssist {
    private static readonly MAX_POPUP_WIDTH = 400;
    private static readonly MAX_POPUP_HEIGHT = 300;

    enabled = true;
    readonly proposalProviders: ContentProposalProvider[] = [];
    private activeTemplateProposal: TemplateProposal | undefined;
    private activeTemplateContext: TemplateContext | undefined;
    private readonly popupContainer: HTMLDivElement;
    private readonly infoPopupContainer: HTMLDivElement;

    constructor(
        private readonly node: HTMLElement,
        private readonly adapter: ContentAssistAdapter,
        private readonly historyManager: HistoryManager
    ) {
        node.addEventListener("keyup", e => {
            if (
                (e.target instanceof HTMLInputElement && e.target.type === "text") ||
                e.target instanceof HTMLTextAreaElement
            )
                return;
            if (e.key === "." || (e.key === " " && e.ctrlKey))
                this.show({});
        });

        this.popupContainer = document.createElement("div");
        this.popupContainer.style.width = DefaultContentAssist.MAX_POPUP_WIDTH + "px";
        this.popupContainer.style.height = DefaultContentAssist.MAX_POPUP_HEIGHT + "px";
        this.popupContainer.style.display = "none";
        this.popupContainer.style.position = "absolute";
        this.popupContainer.style.zIndex = "2000";
        document.body.append(this.popupContainer);

        this.infoPopupContainer = document.createElement("div");
        this.infoPopupContainer.style.display = "none";
        this.infoPopupContainer.style.position = "absolute";
        this.infoPopupContainer.style.maxWidth = DefaultContentAssist.MAX_POPUP_WIDTH + "px";
        this.infoPopupContainer.style.maxHeight = DefaultContentAssist.MAX_POPUP_HEIGHT + "px";
        this.infoPopupContainer.style.zIndex = "2000";
        this.infoPopupContainer.style.overflow = "auto";
        document.body.append(this.infoPopupContainer);

        const focusoutListener = (e: FocusEvent) => {
            if (!e.relatedTarget ||
                (!this.popupContainer.contains(e.relatedTarget as HTMLElement) &&
                    !this.infoPopupContainer.contains(e.relatedTarget as HTMLDivElement)))
                this.handleCancel();
        };
        this.popupContainer.addEventListener("focusout", focusoutListener);
        this.infoPopupContainer.addEventListener("focusout", focusoutListener);
    }

    show(props: ContentAssistProps): void {
        if (!this.enabled)
            return;
        const bounds = this.adapter.getInsertionBounds();
        this.showInfo.cancel();
        this.popupContainer.style.display = "unset";
        this.infoPopupContainer.style.display = "none";
        this.updateBounds(bounds);
        ReactDOM.unmountComponentAtNode(this.popupContainer);
        this.render(props);
    }

    dispose(): void {
        this.popupContainer.remove();
        this.infoPopupContainer.remove();
    }

    private updateBounds(bounds: InsertionBounds) {
        this.popupContainer.style.top = Math.min(window.innerHeight - DefaultContentAssist.MAX_POPUP_HEIGHT, (bounds.y + bounds.height)) + "px";
        this.popupContainer.style.left = Math.min(window.innerWidth - DefaultContentAssist.MAX_POPUP_WIDTH, bounds.x) + "px";
    }

    private handleCancel() {
        this.activeTemplateContext = undefined;
        this.activeTemplateProposal = undefined;
        this.showInfo.cancel();
        this.hideInfo();
        this.popupContainer.style.display = "none";
        ReactDOM.unmountComponentAtNode(this.popupContainer);
        this.node.focus();
    }

    private async handleApply(e: ApplyEvent) {
        const context = {
            historyManager: this.historyManager,
            target: this.adapter.getTarget(),
            selection: this.adapter.getSelection(),
            ...e
        };
        const selection = await e.proposal.apply(context);

        const _activeTemplateProposal = this.activeTemplateProposal;

        this.handleCancel();

        if (selection)
            await this.adapter.setSelection(selection);

        if (e.proposal.postApply)
            e.proposal.postApply(selection || context.selection);

        if (TemplateProposal.is(e.proposal))
            this.activeTemplateProposal = e.proposal;
        else
            this.activeTemplateProposal = _activeTemplateProposal;

        if (this.activeTemplateProposal) {
            this.activeTemplateContext = this.activeTemplateProposal.next(context, [...this.proposalProviders], e.proposal);

            if (this.activeTemplateContext.selection)
                await this.adapter.setSelection(this.activeTemplateContext.selection);

            if (!this.activeTemplateContext.proposalProviders.length) {
                if (this.activeTemplateProposal.postTemplateApply)
                    this.activeTemplateProposal.postTemplateApply();
                this.activeTemplateProposal = undefined;
                this.activeTemplateContext = undefined;
            } else {
                const searchPlaceholder = this.activeTemplateContext.message;
                this.show({
                    searchPlaceholder
                });
            }
        }
    }

    private handleSelectionChange(proposal?: ContentProposal) {
        if (proposal)
            this.showInfo(proposal);
        else
            this.hideInfo();
    }

    private readonly showInfo = debounce(proposal => this.doShowInfo(proposal), 300);

    private doShowInfo(proposal: ContentProposal) {
        if (!("information" in proposal)) {
            this.infoPopupContainer.style.display = "none";
            return;
        }

        this.infoPopupContainer.style.top = this.popupContainer.style.top;
        const popupBounds = this.popupContainer.getBoundingClientRect();
        const maxWidth = document.body.getBoundingClientRect().width;

        if (popupBounds.right + DefaultContentAssist.MAX_POPUP_WIDTH <= maxWidth) {
            this.infoPopupContainer.style.left = popupBounds.right + "px";
            this.infoPopupContainer.style.right = "unset";
            this.infoPopupContainer.style.float = "unset";
        }
        else {
            this.infoPopupContainer.style.right = (maxWidth - popupBounds.left) + "px";
            this.infoPopupContainer.style.left = "unset";
        }
        this.infoPopupContainer.style.display = "unset";
        this.renderInfo(proposal);
    }

    private hideInfo() {
        this.infoPopupContainer.style.display = "none";
        ReactDOM.unmountComponentAtNode(this.infoPopupContainer);
    }

    private renderInfo(proposal: ContentProposal) {
        ReactDOM.render(<InfoPopup proposal={proposal} />, this.infoPopupContainer);
    }

    private render(props: ContentAssistProps) {
        ReactDOM.render(<ContentAssistPopup
            proposalProviders={this.activeTemplateContext ? this.activeTemplateContext.proposalProviders : this.proposalProviders}
            defaultSearch={props.search}
            searchPlaceholder={props.searchPlaceholder}
            onApply={e => this.handleApply(e)}
            onCancel={() => this.handleCancel()}
            onSelectionChange={proposal => this.handleSelectionChange(proposal)}
            context={{
                selection: this.adapter.getSelection(),
                target: this.adapter.getTarget()
            }}
        />, this.popupContainer);
    }
}
