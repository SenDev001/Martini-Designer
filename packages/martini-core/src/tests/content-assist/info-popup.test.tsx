
import * as React from "react";
import * as Renderer from "react-test-renderer";
import { ContentProposal } from "../../browser/content-assist/content-assist";
import { InfoPopup } from "../../browser/content-assist/info-popup";

test("Should render InfoPopup with proposal information", () => {
    const proposal: ContentProposal = {
        apply: jest.fn(),
        label: "Test",
        get information() {
            return <p>Hello</p>;
        }
    };
    const component = Renderer.create(<InfoPopup
        proposal={proposal}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
