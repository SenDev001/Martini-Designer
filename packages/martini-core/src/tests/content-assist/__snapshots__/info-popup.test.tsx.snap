// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render InfoPopup with proposal information 1`] = `
<div
  className="sc-AxjAm hKzUje"
  tabIndex={-1}
>
  <p>
    Hello
  </p>
</div>
`;
